### PR TITLE
Components: improve `vertical` and RTL support for `Divider`

### PR DESCRIPTION
**Pull request by ciampo**
## Description

Following up some feedback from #36302, This PR brings a bunch of improvements to the `Divider` components:

- Added styles for the `vertical` orientation
- Replaced `marginTop` and `marginBottom` props with `marginStart` and `marginEnd` (which are a better abstraction regardless of the orientation). Added RTL support in the process
- Rewrote Storybook examples to feature both Horizontal and Vertical case scenarios
- Removed the `border-color` override to black from Storybook, and changed the default `Divider`'s `border-color` to be `currentColor`
- Fixed/updated documentation
- Added a CHANGELOG entry

## How has this been tested?

Unit tests pass, storybook works as expected

## Types of changes

New feature (vertical styles), Refactor (margin top/bottom => start/end)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
Thanks so much for the improvement. I'll defer to other reviewers for the more technical bits, but as a tool in our toolbox this one seems stronger now. Whether we find a need for the component I'm not yet sure — I think we will as part of a block toolbar refactor.

One question about this part:

```
border-color: #000;
```

I'm always anxious about hardcoded colors, even if they can be updated everywhere by editing a single component. But beyond the hard coded color (which could become a variable), I'm also starting to think about what we'll have to do when we can start bringing dark mode to the editor, how will that be handled? One potentially hacky solution could be to replace the color value with `currentColor`, and make the divider more of a utility than an opinionated separate design. What do you think?

**Comment by ciampo** [comment]
> One question about this part:
> 
> ```
> border-color: #000;
> ```
> 
> I'm always anxious about hardcoded colors, even if they can be updated everywhere by editing a single component. But beyond the hard coded color (which could become a variable), I'm also starting to think about what we'll have to do when we can start bringing dark mode to the editor, how will that be handled? One potentially hacky solution could be to replace the color value with `currentColor`, and make the divider more of a utility than an opinionated separate design. What do you think?

That override of the border color is there specifically for the Storybook example to make the `Divider` easier to spot in the example; it was there before this PR, but it probably creates confusion — should we remove this override and display the `Divider` with its [default color](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/config-values.js#L42)? 

With regards to theming and dark mode, there are many ways to tackle this problem, using a mix of CSS APIs (like CSS variables and the [`prefers-color-scheme`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) media query). I would advise against hacky solutions at this point, and wait until we come up with a more comprehensive strategy around how we implement theming in `@wordpress/components` and Gutenberg

**Comment by jasmussen** [comment]
> should we remove this override and display the Divider with its default color?

I think so, but I also think its default color should be currentColor and not a semitransparent black. [Those defaults](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/config-values.js#L42) in general feel worth looking at (separately), as none of them feel very compatible with the design language of the block editor (for which there are some decent examples in #34574). To an extent, those defaults should look a bit more like those defined in our base styles: [colors](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_colors.scss), [other metrics](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_variables.scss).

> With regards to theming and dark mode, there are many ways to tackle this problem, using a mix of CSS APIs (like CSS variables and the prefers-color-scheme media query). I would advise against hacky solutions at this point, and wait until we come up with a more comprehensive strategy around how we implement theming in @wordpress/components and Gutenberg

👌

**Comment by ciampo** [comment]
Thanks for the feedback, @jasmussen ! I've updated this PR and followed your advice:

- We don't override the border-color anymore in Storybook 
- The default border-color is `currentColor`

**Comment by ciampo** [comment]
This PR is ready for a final review — @mirka and @diegohaz , do you mind taking a look ? Thank you!

File: packages/components/src/divider/component.tsx
Comment (mirka) [inline packages/components/src/divider/component.tsx]: Am I missing something, or is the logic in this function overly convoluted?

For example, would something simpler like this capture the intent better? 👇

```js
return {
  marginStart: marginStart ?? marginTop,
  marginEnd: marginEnd ?? marginBottom,
  ...otherProps,
};
```

(If there are more sensitive requirements that were intended, e.g. specifically differentiating zero/null/undefined values, it may be better to add comments or unit tests because it's a little hard to tell.)

File: packages/components/src/divider/stories/index.js
Comment (mirka) [inline packages/components/src/divider/stories/index.js]: Would this be a good time to refactor these Knobs into Controls? 😄

File: packages/components/src/divider/styles.ts
Comment (mirka) [inline packages/components/src/divider/styles.ts:32]: TIL `aria-orientation`!

**Review by mirka — COMMENTED** [review]
Really nice improvements, logical props ftw 💥 

Maybe not in scope for this PR, but I noticed that this component doesn't include a CSS reset for the default borders on the `hr` element. It's intended to be just a 1px separator, correct?

File: packages/components/src/divider/styles.ts
Comment (ciampo) [reply packages/components/src/divider/styles.ts:32]: That's thanks to [Reakit](https://reakit.io/docs/separator/) !

File: packages/components/src/divider/stories/index.js
Comment (ciampo) [reply packages/components/src/divider/stories/index.js]: I can definitely look into it!

File: packages/components/src/divider/styles.ts
Comment (ciampo) [inline packages/components/src/divider/styles.ts:30]: > I noticed that this component doesn't include a CSS reset for the default borders on the hr element. It's intended to be just a 1px separator, correct?

By resetting, do you mean setting `border: 0` for the `hr` ?

In that case, I may have introduced that regression, because previously the code was overriding the whole `border-width` property, while after my changes, only the `border-[top|right]` props are overridden. Happy to add an additional `border: 0` CSS rule back

File: packages/components/src/divider/component.tsx
Comment (ciampo) [reply packages/components/src/divider/component.tsx]: I initially came up with this pattern on the `Card` component in order to avoid overriding any potential props passed via Context: https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/card/card/hook.js#L35

Probably safer to keep using this pattern to future-proof `Divider`, what do you think?  (in that case, I will definitely add some comments as you suggested)

Comment (ciampo) [reply packages/components/src/divider/component.tsx]: Now that I'm thinking about it, this is an experimental component — and therefore another option could be to just remove support for `marginTop` and `marginBottom` (avoiding `useDeprecatedProps` altogether)

File: packages/components/src/divider/styles.ts
Comment (mirka) [reply packages/components/src/divider/styles.ts:30]: Oh good then 😄 Yes I meant `border: 0` but I didn't realize it was a regression!

File: packages/components/src/divider/component.tsx
Comment (mirka) [reply packages/components/src/divider/component.tsx]: > avoid overriding any potential props passed via Context

Ah that's good to know! 😮 Do you think it's worth addressing this in the `useContextSystem` function itself so it's smarter about undefined properties? It's pretty easy to mess up if you don't know or forget.

> another option could be to just remove support

That's perfectly reasonable!

Comment (ciampo) [reply packages/components/src/divider/component.tsx]: > Ah that's good to know! 😮 Do you think it's worth addressing this in the useContextSystem function itself so it's smarter about undefined properties? It's pretty easy to mess up if you don't know or forget.

I'm not sure, but it's definitely something to look into

> That's perfectly reasonable!

That's my preferred option

Comment (ciampo) [reply packages/components/src/divider/component.tsx]: Done in https://github.com/WordPress/gutenberg/pull/36579/commits/0f6000120dde2cb05dc6417435b977d8206ea1a5

File: packages/components/src/divider/stories/index.js
Comment (ciampo) [reply packages/components/src/divider/stories/index.js]: Done in https://github.com/WordPress/gutenberg/pull/36579/commits/3f0ba5b995342d1bd6e9ca04ad4649f1eeaad19e

File: packages/components/src/divider/styles.ts
Comment (ciampo) [reply packages/components/src/divider/styles.ts:30]: Done in https://github.com/WordPress/gutenberg/pull/36579/commits/d5d44f0d92106f6fed66c0dce7f705299a7a8b91

**Comment by ciampo** [comment]
Hey @mirka , your initial feedback should have all been addressed! I also rebased on top of `trunk` to fix a conflict.

**Review by mirka — APPROVED** [review]
Solid improvements 🚢

**Comment by ciampo** [comment]
Will merge once the PHP unit tests won't fail 😅  (unrelated to this PR)

**Pull request merged**
